File: 240120/숫자 카운트/numeric-count.py
N = int(input())
predicts = [list(map(int, input().split())) for _ in range(N)]
# 서로다른 세자리..100 ~ 999 밖에 없음...
# 전수조사 쌉가능하다..
# 10개 질문 * 1000개의 숫자 * 3자리 > 완탐 충분히 가능!

def check(candidate_number: int, predict: list[int]):
    candidate = []
    pre = []
    for i in range(1, 4):
        a = candidate_number % (10 ** i) // (10 ** (i - 1))
        p = predict[0] % (10 ** i) // (10 ** (i - 1))

        if a in candidate or a == 0:
            return False
        candidate.append(a)

        if p in pre or p == 0:
            return False
        pre.append(p)


    cnt_1 = 0
    cnt_2 = 0

    for a, p in zip(candidate, pre):
        if a == p:
            cnt_1 += 1
        else:
            if p in candidate:
                cnt_2 += 1

    if cnt_1 == predict[1] and cnt_2 == predict[2]:
        return True
    return False


cnt = set(range(100, 1000))
for predict in predicts:
    candidate_list = set()
    for candidate_number in range(100, 1000):
        if check(candidate_number, predict):
            candidate_list.add(candidate_number)   
    # 각 질문에 대해서 교집합으로 만족하는 것만 남기기     
    cnt = cnt & candidate_list


print(len(cnt))

# 틀린이유1.
# 각 질문들이 필터 역할을 해야함, 지금처럼 한 질문에 대해서 100 ~ 1000 전수조사를 때리면 각 질문에 대해서는 만족하는 후보값이
# 얻어지겠지만 모두 만족하는 값이 되는건 아니다.
# 질문들 모두 만족해야함!

# 틀린이유2.
# 서로다른 숫자! 때문에 틀린듯!
# 0도 안포함됨. 서로다른 1~9!

File: 240120/숫자 카운트/numeric-count.java
import java.util.*;

public class Main {

    private static final int MAX_DIGIT = 3;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();

        Set<Integer> cnt = new HashSet<>();
        for (int i = 100; i < 1000; i++) {
            cnt.add(i);
        }

        int[][] predicts = new int[N][3];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < 3; j++) {
                predicts[i][j] = scanner.nextInt();
            }
        }


        for (int[] predict : predicts) {
            Set<Integer> candidateList = new HashSet<>();
            for (int candidateNumber = 100; candidateNumber < 1000; candidateNumber++) {
                //  틀린이유: predict가 변경되고 있음! -> check 함수 변경해야함
                if (check(candidateNumber, predict)) {
                    candidateList.add(candidateNumber);
                }
            }
            // 교집합
            cnt.retainAll(candidateList);
        }

        System.out.println(cnt.size());
    }

    private static boolean check(int candidateNumber, int[] predict) {
        List<Integer> candidate = new ArrayList<>();
        List<Integer> pre = new ArrayList<>();

        for (int i = 1; i <= MAX_DIGIT; i++) {
            // 기존 값이 변해서는 안된다!!
            // candidateNumber % 10
            // candidate / 10 하면 기존 값이 변경된다!
            int a = candidateNumber % (int) Math.pow(10, i) / (int) Math.pow(10, i - 1);
            int p = predict[0] % (int) Math.pow(10, i) / (int) Math.pow(10, i - 1);

            if (candidate.contains(a) || a == 0) {
                return false;
            }
            candidate.add(a);

            if (pre.contains(p) || p == 0) {
                return false;
            }
            pre.add(p);
        }

        int cnt_1 = 0;
        int cnt_2 = 0;

        for (int i = 0; i < MAX_DIGIT; i++) {
            if (candidate.get(i).equals(pre.get(i))) {
                cnt_1++;
            } else {
                if (pre.contains(candidate.get(i))) {
                    cnt_2++;
                }
            }
        }

        return cnt_1 == predict[1] && cnt_2 == predict[2];
    }

}